File: bumi/src/string-matchers.ts
import { RE2JS } from "re2js";

import { MAX_PATTERN_LENGTH, MAX_REGEX_PATTERN_LENGTH, PatternTooLongError } from "./limits.ts";

/**
 * Enforce the literal pattern limit at construction.
 *
 * The limit belongs to the type that holds the pattern, not to the config
 * loader — before 2026-08-17 only the loader checked, so the gateway, direct
 * construction, and the playground's graph renderer were all unguarded.
 * See `DECISIONS.md` D-029.
 */
function checkLiteralLength(pattern: string): void {
	if (pattern.length > MAX_PATTERN_LENGTH) {
		throw new PatternTooLongError(pattern.length, MAX_PATTERN_LENGTH);
	}
}
import { MatcherError } from "./matcher.ts";
import { assertRepeatBudget } from "./regex-budget.ts";
import type { MatchingData } from "./types.ts";

/** Exact string equality. Pre-lowercases at construction when ignore_case. */
export class ExactMatcher {
	private readonly cmpValue: string;

	constructor(
		readonly value: string,
		readonly ignoreCase: boolean = false,
	) {
		checkLiteralLength(value);
		this.cmpValue = ignoreCase ? value.toLowerCase() : value;
	}

	matches(value: MatchingData): boolean {
		if (typeof value !== "string") return false;
		const input = this.ignoreCase ? value.toLowerCase() : value;
		return input === this.cmpValue;
	}
}

/** String prefix match. Pre-lowercases at construction when ignore_case. */
export class PrefixMatcher {
	private readonly cmpPrefix: string;

	constructor(
		readonly prefix: string,
		readonly ignoreCase: boolean = false,
	) {
		checkLiteralLength(prefix);
		this.cmpPrefix = ignoreCase ? prefix.toLowerCase() : prefix;
	}

	matches(value: MatchingData): boolean {
		if (typeof value !== "string") return false;
		const input = this.ignoreCase ? value.toLowerCase() : value;
		return input.startsWith(this.cmpPrefix);
	}
}

/** String suffix match. Pre-lowercases at construction when ignore_case. */
export class SuffixMatcher {
	private readonly cmpSuffix: string;

	constructor(
		readonly suffix: string,
		readonly ignoreCase: boolean = false,
	) {
		checkLiteralLength(suffix);
		this.cmpSuffix = ignoreCase ? suffix.toLowerCase() : suffix;
	}

	matches(value: MatchingData): boolean {
		if (typeof value !== "string") return false;
		const input = this.ignoreCase ? value.toLowerCase() : value;
		return input.endsWith(this.cmpSuffix);
	}
}

/** Substring containment. Pre-lowercases pattern at construction when ignore_case. */
export class ContainsMatcher {
	private readonly cmpSubstring: string;

	constructor(
		readonly substring: string,
		readonly ignoreCase: boolean = false,
	) {
		checkLiteralLength(substring);
		this.cmpSubstring = ignoreCase ? substring.toLowerCase() : substring;
	}

	matches(value: MatchingData): boolean {
		if (typeof value !== "string") return false;
		const input = this.ignoreCase ? value.toLowerCase() : value;
		return input.includes(this.cmpSubstring);
	}
}

/**
 * Regular expression match using RE2 for guaranteed linear-time matching.
 * Uses RE2JS.compile().matcher().find() which searches anywhere in the string
 * (equivalent to Python's re.search, Rust's regex find).
 *
 * RE2 does not support backreferences or lookahead/lookbehind because they
 * require backtracking. Patterns using them are rejected at compile time.
 */
export class RegexMatcher {
	private readonly compiled: RE2JS;

	/**
	 * @throws {PatternTooLongError} if the pattern exceeds the length limit
	 * @throws {MatcherError} if the repetition budget is exceeded, or the
	 *   pattern does not compile
	 *
	 * Both limits are enforced **here**, in the constructor that owns the
	 * compiled program, not in the config loader. Every caller inherits them —
	 * the registry, the gateway, and the playground's graph renderer, which
	 * calls `parseMatcherConfig` without `loadMatcher` and so previously
	 * inherited nothing. See D-029.
	 */
	constructor(readonly pattern: string) {
		if (pattern.length > MAX_REGEX_PATTERN_LENGTH) {
			throw new PatternTooLongError(pattern.length, MAX_REGEX_PATTERN_LENGTH);
		}
		// re2js supplies no compile-time budget of its own — see regex-budget.ts.
		try {
			assertRepeatBudget(pattern);
		} catch (e) {
			throw new MatcherError(
				`invalid regex pattern "${pattern}": ${e instanceof Error ? e.message : String(e)}`,
			);
		}
		try {
			this.compiled = RE2JS.compile(pattern);
		} catch (e) {
			throw new MatcherError(
				`invalid regex pattern "${pattern}": ${e instanceof Error ? e.message : String(e)}`,
			);
		}
	}

	matches(value: MatchingData): boolean {
		if (typeof value !== "string") return false;
		return this.compiled.matcher(value).find();
	}
}

File: puma/src/xuma/_string_matchers.py
"""Concrete string matchers implementing the InputMatcher protocol.

Each matcher is a frozen dataclass — immutable after construction.
All matchers return False for non-string or None input values.

Regex uses ``google-re2`` for guaranteed linear-time matching. RE2 does not
support backreferences or lookahead/lookbehind because they require
backtracking — patterns using them are rejected at compile time.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

import re2

from xuma._limits import MAX_PATTERN_LENGTH, MAX_REGEX_PATTERN_LENGTH
from xuma._matcher import MatcherError

if TYPE_CHECKING:
    from xuma._types import MatchingData


def _check_literal_length(pattern: str) -> None:
    """Enforce the literal pattern limit at construction.

    The limit belongs to the type that holds the pattern, not to the config
    loader -- until 2026-08-17 only the loader checked, so the HTTP gateway and
    direct construction were unguarded.
    """
    if not isinstance(pattern, str):
        msg = f"pattern must be str, got {type(pattern).__name__}"
        raise MatcherError(msg)
    if len(pattern) > MAX_PATTERN_LENGTH:
        msg = f"pattern length {len(pattern)} exceeds maximum {MAX_PATTERN_LENGTH}"
        raise MatcherError(msg)


@dataclass(frozen=True, slots=True)
class ExactMatcher:
    """Exact string equality match.

    When ignore_case is True, comparison is case-insensitive.
    The comparison value is pre-lowercased at construction time.
    """

    value: str
    ignore_case: bool = False
    _cmp_value: str = field(init=False, repr=False)

    def __post_init__(self) -> None:
        _check_literal_length(self.value)
        object.__setattr__(
            self, "_cmp_value", self.value.casefold() if self.ignore_case else self.value
        )

    def matches(self, value: MatchingData, /) -> bool:
        if not isinstance(value, str):
            return False
        input_val = value.casefold() if self.ignore_case else value
        return input_val == self._cmp_value


@dataclass(frozen=True, slots=True)
class PrefixMatcher:
    """String prefix match (startswith).

    When ignore_case is True, comparison is case-insensitive.
    The prefix is pre-lowercased at construction time.
    """

    prefix: str
    ignore_case: bool = False
    _cmp_prefix: str = field(init=False, repr=False)

    def __post_init__(self) -> None:
        _check_literal_length(self.prefix)
        object.__setattr__(
            self, "_cmp_prefix", self.prefix.casefold() if self.ignore_case else self.prefix
        )

    def matches(self, value: MatchingData, /) -> bool:
        if not isinstance(value, str):
            return False
        input_val = value.casefold() if self.ignore_case else value
        return input_val.startswith(self._cmp_prefix)


@dataclass(frozen=True, slots=True)
class SuffixMatcher:
    """String suffix match (endswith).

    When ignore_case is True, comparison is case-insensitive.
    The suffix is pre-lowercased at construction time.
    """

    suffix: str
    ignore_case: bool = False
    _cmp_suffix: str = field(init=False, repr=False)

    def __post_init__(self) -> None:
        _check_literal_length(self.suffix)
        object.__setattr__(
            self, "_cmp_suffix", self.suffix.casefold() if self.ignore_case else self.suffix
        )

    def matches(self, value: MatchingData, /) -> bool:
        if not isinstance(value, str):
            return False
        input_val = value.casefold() if self.ignore_case else value
        return input_val.endswith(self._cmp_suffix)


@dataclass(frozen=True, slots=True)
class ContainsMatcher:
    """Substring search match.

    When ignore_case is True, comparison is case-insensitive.
    The substring pattern is pre-lowercased at construction time
    (Knuth optimization: avoid repeated lowercasing of the pattern).
    """

    substring: str
    ignore_case: bool = False
    _cmp_substring: str = field(init=False, repr=False)

    def __post_init__(self) -> None:
        _check_literal_length(self.substring)
        object.__setattr__(
            self,
            "_cmp_substring",
            self.substring.casefold() if self.ignore_case else self.substring,
        )

    def matches(self, value: MatchingData, /) -> bool:
        if not isinstance(value, str):
            return False
        input_val = value.casefold() if self.ignore_case else value
        return self._cmp_substring in input_val


@dataclass(frozen=True, slots=True)
class RegexMatcher:
    """Regular expression match.

    The pattern is compiled at construction time via ``google-re2``, providing
    guaranteed linear-time matching. Uses search (not fullmatch) to match
    anywhere in the string, consistent with rumi's behavior.

    RE2 does not support backreferences or lookahead/lookbehind because they
    require backtracking. Patterns using them are rejected at compile time.

    Raises:
        MatcherError: If the pattern is not valid RE2 syntax.
    """

    pattern: str
    _compiled: re2.Pattern[str] = field(init=False, repr=False)

    def __post_init__(self) -> None:
        # The limit is enforced here, in the constructor that owns the compiled
        # program, not in the config loader. Every caller inherits it -- the
        # registry, the HTTP gateway, and direct construction. Until 2026-08-17
        # the loader was the only guarded path, so the gateway accepted a
        # 40,960-byte regex against this 4,096 limit.
        if not isinstance(self.pattern, str):
            # google-re2 raises a bare TypeError for a non-str pattern, which
            # escapes the MatcherError contract callers rely on (review L-4).
            msg = f"regex pattern must be str, got {type(self.pattern).__name__}"
            raise MatcherError(msg)
        if len(self.pattern) > MAX_REGEX_PATTERN_LENGTH:
            msg = (
                f"pattern length {len(self.pattern)} exceeds maximum "
                f"{MAX_REGEX_PATTERN_LENGTH}"
            )
            raise MatcherError(msg)
        try:
            compiled = re2.compile(self.pattern)
        except re2.error as e:
            msg = f'invalid regex pattern "{self.pattern}": {e}'
            raise MatcherError(msg) from e
        except TypeError as e:
            # google-re2 raises TypeError for a non-str pattern, which escapes
            # the MatcherError contract callers rely on (review L-4).
            msg = f"invalid regex pattern {self.pattern!r}: {e}"
            raise MatcherError(msg) from e
        object.__setattr__(self, "_compiled", compiled)

    def matches(self, value: MatchingData, /) -> bool:
        if not isinstance(value, str):
            return False
        return self._compiled.search(value) is not None


@dataclass(frozen=True, slots=True)
class BoolMatcher:
    """Matches a boolean value.

    Reachable through ``custom_match``, which is the seam for comparisons xDS's
    own ``StringMatcher`` cannot express — a boolean is one, since
    ``value_match`` only compares strings.

    No input x.uma ships produces a boolean; this exists for a custom domain
    whose ``DataInput`` returns one. It is registered by
    :func:`~xuma._registry.register_core_matchers` so that
    ``xuma.core.v1.BoolMatcher`` resolves in all implementations rather than in
    some — that function previously registered nothing at all while its
    docstring said it did.
    """

    expected: bool

    def matches(self, value: MatchingData, /) -> bool:
        return isinstance(value, bool) and value == self.expected
